crimson/osd/recovery_backend: always set the primary osd's BackfillInterval::version to the last_update before the backfill scan

xxhdx1985126 opened the pull request:
Fixes: https://tracker.ceph.com/issues/67916
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

xxhdx1985126 commented:
This PR is based on https://github.com/ceph/ceph/pull/58868

athanatos commented:
If I understood https://tracker.ceph.com/issues/63844 correctly, storing a copy of the obs before mutation would also fix the problem -- I think that solution would be a lot cleaner than rereading the obc on error.

xxhdx1985126 commented:
> If I understood https://tracker.ceph.com/issues/63844 correctly, storing a copy of the obs before mutation would also fix the problem -- I think that solution would be a lot cleaner than rereading the obc on error.

Yeah, I've changed the relative code and moved the commit to https://github.com/ceph/ceph/pull/59543, please take a look again.

Matan-B commented:
https://github.com/ceph/ceph/pull/59543#pullrequestreview-2288497152

Matan-B commented on src/crimson/osd/recovery_backend.cc:
```diff
@@ -219,60 +219,59 @@ RecoveryBackend::handle_backfill_remove(
-	crimson::osd::ObjectContextRef obc;
-	if (pg.is_primary()) {
-	  obc = pg.obc_registry.maybe_get_cached_obc(object);
+  try {
```
As discussed offline,  we can avoid try/catch usage here

xxhdx1985126 replied:
Done, please take a look again, thanks:-)

Matan-B reviewed (commented):
If possible, can you please try dissecting the commit into smaller logical diffs? It could help a bit with reviewing

Matan-B reviewed (commented):
Can we re-use our stages mechanism for this purpose?
We should be able to make client requests mutually exclusive from backfill scans by synchronizing between the two with a OrderedExclusivePhaseT, right?

Note: This won't be possible with the current implementation suggested in #60185 where backfill events are not synchronized with PeeringEvent.

athanatos commented:
From the bug, it looks like the thing we're trying to solve here is:
1. write foo
2. backfill scan interval including foo
3. foo commits

Step 2 here wouldn't contain foo as the operation hadn't necessary committed yet.

We really don't want to make backfill scan and client IO mutually exclusive -- it would mean that either client writes are blocked until the scan completes or we allow client IO to preempt and prevent backfill progress.  Neither option is attractive.

Fortunately, we know what writes have happened recently as they are in the pg log.  From PrimaryLogPG.cc:

```
/**
 * recover_backfill
 *
 * Invariants:
 *
 * backfilled: fully pushed to replica or present in replica's missing set (both
 * our copy and theirs).
 *
 * All objects on a backfill_target in
 * [MIN,peer_backfill_info[backfill_target].begin) are valid; logically-removed
 * objects have been actually deleted and all logically-valid objects are replicated.
 * There may be PG objects in this interval yet to be backfilled.
 *
 * All objects in PG in [MIN,backfill_info.begin) have been backfilled to all
 * backfill_targets.  There may be objects on backfill_target(s) yet to be deleted.
 *
 * For a backfill target, all objects < std::min(peer_backfill_info[target].begin,
 *     backfill_info.begin) in PG are backfilled.  No deleted objects in this
 * interval remain on the backfill target.
 *
 * For a backfill target, all objects <= peer_info[target].last_backfill
 * have been backfilled to target
 *
 * There *MAY* be missing/outdated objects between last_backfill_started and
 * std::min(peer_backfill_info[*].begin, backfill_info.begin) in the event that client
 * io created objects since the last scan.  For this reason, we call
 * update_range() again before continuing backfill.
 */
uint64_t PrimaryLogPG::recover_backfill(
  uint64_t max,
  ThreadPool::TPHandle &handle, bool *work_started)
```

That last paragraph in particular is the relevant bit:

```
 * There *MAY* be missing/outdated objects between last_backfill_started and
 * std::min(peer_backfill_info[*].begin, backfill_info.begin) in the event that client
 * io created objects since the last scan.  For this reason, we call
 * update_range() again before continuing backfill.
```

The idea is that when we go to actually use the interval to start pushes, we scan the pg log for any events that happened between the start of the scan and now to bring the interval up to date.  BackfillState does seem to have logic for this:

```
BackfillState::Enqueuing::Enqueuing(my_context ctx)
  : my_base(ctx)
{
  auto& primary_bi = backfill_state().backfill_info;

  // update our local interval to cope with recent changes
  primary_bi.begin = backfill_state().last_backfill_started;
  if (primary_bi.version < peering_state().get_log_tail()) {
    // it might be that the OSD is so flooded with modifying operations
    // that backfill will be spinning here over and over. For the sake
    // of performance and complexity we don't synchronize with entire PG.
    // similar can happen in classical OSD.
    logger().warn("{}: bi is old, rescanning of local backfill_info",
                  __func__);
    post_event(RequestPrimaryScanning{});
    return;
  } else {
    maybe_update_range();
  }
```

I suspect that the bug is likely related to that mechanism.

xxhdx1985126 commented:
> From the bug, it looks like the thing we're trying to solve here is:
> 
> 1. write foo
> 2. backfill scan interval including foo
> 3. foo commits
> 
> Step 2 here wouldn't contain foo as the operation hadn't necessary committed yet.
> 
> We really don't want to make backfill scan and client IO mutually exclusive -- it would mean that either client writes are blocked until the scan completes or we allow client IO to preempt and prevent backfill progress. Neither option is attractive.
> 
> Fortunately, we know what writes have happened recently as they are in the pg log. From PrimaryLogPG.cc:
> 
> ```
> /**
>  * recover_backfill
>  *
>  * Invariants:
>  *
>  * backfilled: fully pushed to replica or present in replica's missing set (both
>  * our copy and theirs).
>  *
>  * All objects on a backfill_target in
>  * [MIN,peer_backfill_info[backfill_target].begin) are valid; logically-removed
>  * objects have been actually deleted and all logically-valid objects are replicated.
>  * There may be PG objects in this interval yet to be backfilled.
>  *
>  * All objects in PG in [MIN,backfill_info.begin) have been backfilled to all
>  * backfill_targets.  There may be objects on backfill_target(s) yet to be deleted.
>  *
>  * For a backfill target, all objects < std::min(peer_backfill_info[target].begin,
>  *     backfill_info.begin) in PG are backfilled.  No deleted objects in this
>  * interval remain on the backfill target.
>  *
>  * For a backfill target, all objects <= peer_info[target].last_backfill
>  * have been backfilled to target
>  *
>  * There *MAY* be missing/outdated objects between last_backfill_started and
>  * std::min(peer_backfill_info[*].begin, backfill_info.begin) in the event that client
>  * io created objects since the last scan.  For this reason, we call
>  * update_range() again before continuing backfill.
>  */
> uint64_t PrimaryLogPG::recover_backfill(
>   uint64_t max,
>   ThreadPool::TPHandle &handle, bool *work_started)
> ```
> 
> That last paragraph in particular is the relevant bit:
> 
> ```
>  * There *MAY* be missing/outdated objects between last_backfill_started and
>  * std::min(peer_backfill_info[*].begin, backfill_info.begin) in the event that client
>  * io created objects since the last scan.  For this reason, we call
>  * update_range() again before continuing backfill.
> ```
> 
> The idea is that when we go to actually use the interval to start pushes, we scan the pg log for any events that happened between the start of the scan and now to bring the interval up to date. BackfillState does seem to have logic for this:
> 
> ```
> BackfillState::Enqueuing::Enqueuing(my_context ctx)
>   : my_base(ctx)
> {
>   auto& primary_bi = backfill_state().backfill_info;
> 
>   // update our local interval to cope with recent changes
>   primary_bi.begin = backfill_state().last_backfill_started;
>   if (primary_bi.version < peering_state().get_log_tail()) {
>     // it might be that the OSD is so flooded with modifying operations
>     // that backfill will be spinning here over and over. For the sake
>     // of performance and complexity we don't synchronize with entire PG.
>     // similar can happen in classical OSD.
>     logger().warn("{}: bi is old, rescanning of local backfill_info",
>                   __func__);
>     post_event(RequestPrimaryScanning{});
>     return;
>   } else {
>     maybe_update_range();
>   }
> ```
> 
> I suspect that the bug is likely related to that mechanism.

Ah, yes, I probably misunderstood the situation.

After a scan of the code, I think the issue is probably as follows:

1. At the end of `RecoveryBackend::scan_for_backfill()`, backfill interval's version is updated to `peering_state.info.last_update`;
2. After the `io` is submitted to `SeaStore`, `peering_state.info.last_update` is updated to `PG::projected_last_update` no matter whether the `io` is committed;

So, the backfill interval's version may be updated to `PG::projected_last_update` even it doesn't include the latest change, which makes later `BackfillState::maybe_update_range()` skip the most recent changes. @athanatos Am I right?

If so, I think the most straightforward fix is to copy `peering_state.info.last_update` at the beginning of `RecoveryBackend::scan_for_backfill()` and assign it to the backfill interval at the end. @athanatos What do you think?

xxhdx1985126 commented:
> Can we re-use our stages mechanism for this purpose? We should be able to make client requests mutually exclusive from backfill scans by synchronizing between the two with a OrderedExclusivePhaseT, right?
> 
> Note: This won't be possible with the current implementation suggested in #60185 where backfill events are not synchronized with PeeringEvent.

Sorry, my previous analysis of the issue was probably wrong. If I'm right this time(https://github.com/ceph/ceph/pull/59615#issuecomment-2443117396), we don't have to do any synchronization at all.

Matan-B commented:
> Ah, yes, I probably misunderstood the situation.
> 
> After a scan of the code, I think the issue is probably as follows:
> 
> 1. At the end of `RecoveryBackend::scan_for_backfill()`, backfill interval's version is updated to `peering_state.info.last_update`;
> 2. After the `io` is submitted to `SeaStore`, `peering_state.info.last_update` is updated to `PG::projected_last_update` no matter whether the `io` is committed;
> 
> So, the backfill interval's version may be updated to `PG::projected_last_update` even it doesn't include the latest change, which makes later `BackfillState::maybe_update_range()` skip the most recent changes. @athanatos Am I right?
> 
> If so, I think the most straightforward fix is to copy `peering_state.info.last_update` at the beginning of `RecoveryBackend::scan_for_backfill()` and assign it to the backfill interval at the end. @athanatos @Matan-B What do you think?

Can we upgrade `maybe_update_range()` to support scanning the `projected_log` as well and update the interval again before start pushing?

athanatos commented:
Right, you want to capture last_update at the the time of the start of the scan because any update <= last_update is reflected in the local store (note, the primary and each replica tag their own interval with their own *local* last_update -- they won't necessarily match).  @Matan-B is also correct, when you scan the log, you need to scan the projected log as well.  See PrimaryLogPG::update_range.

xxhdx1985126 commented:
> Right, you want to capture last_update at the the time of the start of the scan because any update <= last_update is reflected in the local store (note, the primary and each replica tag their own interval with their own _local_ last_update -- they won't necessarily match). @Matan-B is also correct, when you scan the log, you need to scan the projected log as well. See PrimaryLogPG::update_range.

Yeah, but it seems that crimson hasn't supported projected_log yet, maybe we can fall back to check on `peering_state.info.last_update`, instead of `PG::projected_last_update` and scan the pg log in `BackfillState::Enqueuing::maybe_update_range()` to include new modifications for now, so that we can quickly fix this issue. In the future, when crimson does support projected log, we change `BackfillState::Enqueuing::maybe_update_range()` to check on the projected_last_update again. What do you think? @athanatos @Matan-B

Matan-B commented:
> Yeah, but it seems that crimson hasn't supported projected_log yet, maybe we can fall back to check on `peering_state.info.last_update`, instead of `PG::projected_last_update`, and scan the pg log in `BackfillState::Enqueuing::maybe_update_range()` to include new modifications for now, so that we can quickly fix this issue. In the future, when crimson does support projected log, we change `BackfillState::Enqueuing::maybe_update_range()` to check on the projected_last_update again. What do you think? @athanatos @Matan-B

https://github.com/ceph/ceph/pull/56832 projected log should be ready here.
If something else is needed, you can take the commit from this PR or if you could review it so I'll test it.

xxhdx1985126 commented:
> > Yeah, but it seems that crimson hasn't supported projected_log yet, maybe we can fall back to check on `peering_state.info.last_update`, instead of `PG::projected_last_update`, and scan the pg log in `BackfillState::Enqueuing::maybe_update_range()` to include new modifications for now, so that we can quickly fix this issue. In the future, when crimson does support projected log, we change `BackfillState::Enqueuing::maybe_update_range()` to check on the projected_last_update again. What do you think? @athanatos @Matan-B
> 
> #56832 projected log should be ready here. If something else is needed, you can take the commit from this PR or if you could review it so I'll test it.

Just approve it:-)

Matan-B commented on src/crimson/osd/recovery_backend.cc:
```diff
@@ -265,7 +265,6 @@ RecoveryBackend::scan_for_backfill(
       BackfillInterval bi;
       bi.begin = std::move(start);
       bi.end = std::move(next);
-      bi.version = pg.get_info().last_update;
```
Do we still need this change if we will support projected log? I would prefer updating the interval version when we actually populate it so that the version will make sense.

btw, I've pushed a refactor to this code since it was quite confusing and there was a bug hidden there - https://github.com/ceph/ceph/pull/60522

xxhdx1985126 replied:
Yes, I think we still need this, otherwise, backfill interval's version may be ser to a relative new value, and skip some new modifications.

Matan-B replied:
I'm not sure I agree with this change (or perhaps not understanding it) , `version` is specifically there to know the version in which the scan has **occurred**.  

From the earlier discussion with @athanatos:
>Right, you want to capture last_update at the the time of the start of the scan because any update <= last_update is reflected in the local store (note, the primary and each replica tag their own interval with their own local last_update -- they won't necessarily match)

With the change here we will be capturing `last_update` after completing the scan while reacting to `PrimaryScanned`. Should we instead capture the version **before** the scan, in `PrimaryScanning::PrimaryScanning(my_context ctx)`?
Which as I can see we already are. We seem to overwrite the version  captured here (in `scan_for_backfill`) and this may have caused the issue?

```
// -- PrimaryScanning
BackfillState::PrimaryScanning::PrimaryScanning(my_context ctx)
  : my_base(ctx)
{
  backfill_state().backfill_info.version = peering_state().get_last_update(); <---
  backfill_listener().request_primary_scan(
    backfill_state().backfill_info.begin);
}
```

Seems like classic does the same thing by setting the version **before** scanning:
```
    bi->version = info.last_update;
    scan_range(local_min, local_max, bi, handle);
```

xxhdx1985126 replied:
> With the change here we will be capturing `last_update` after completing the scan while reacting to `PrimaryScanned`. Should we instead capture the version **before** the scan, in `PrimaryScanning::PrimaryScanning(my_context ctx)`

`PrimaryScanning::PrimaryScanning(my_context ctx)` already captures `peering_state.info.last_update`, see the following code: https://github.com/ceph/ceph/blob/de3c6e4ea1488028dfb66c20c23cb938b23cac88/src/crimson/osd/backfill_state.cc#L409C1-L409C78

Matan-B replied:
Ack, I'm planning to refactor a bit the way we handle interval scans and versions updates to make it easier to follow.

Matan-B commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -394,6 +394,7 @@ boost::statechart::result
 BackfillState::PrimaryScanning::react(PrimaryScanned evt)
 {
   logger().debug("{}", __func__);
+  evt.result.version = backfill_state().backfill_info.version;
```
Continuing my previous comment, I think that this will be correct without this line.

xxhdx1985126 replied:
We are not capturing `peering_state.info.last_update` here, at this moment `backfill_state().backfill_info.version` is the last_update captured in `PrimaryScanning::PrimaryScanning(my_context ctx)`

xxhdx1985126 replied:
Without this line, `backfill_state().backfill_info.version` will be overwritten with the empty value.

Matan-B replied:
Got it, approving. I'll add a comment to this section in my refactor to make it more explicit. I plan to require passing version once populating the Interval. Thanks for explaining

xxhdx1985126 replied:
@Matan-B Sorry, I think I was wrong and you were correct, we should only set the primary osd's backfill_info.version after we've done "update_range" when we can be sure we have all objects before `projected_last_update`, there's no point for us to set it before "update_range". @athanatos @Matan-B Am I right? If so, I'll close this PR.

xxhdx1985126 replied:
@Matan-B This line is wrong, you're right.

xxhdx1985126 replied:
@athanatos @Matan-B Maybe always setting `primary_bi.version` to `last_complete_ondisk` can solve this issue, what do you think?

Matan-B replied:
>1. An IO request set `projected_last_update` to `100'200` and creates object `obj_k`;
>2. before the IO request is committed, a primary scan happens, adds all objects from `obj_a to obj_h` to the backfill set and set `primary_bi.version` to `100'200`
>3. also before the IO request is committed, another round of primary scan happens and adds all objects from `obj_i to obj_m` to the backfill set, at which point, `obj_k` won't be in the backfill set because the IO request is not committed and `primary_bi.version` is already at `100'200`
>4. `obj_k` won't be backfilled eventually.

Once  `projected_last_update` is set to `100'200` (step 1), and `primary_bi.version` is set to `100'200` (step2) - that means that the `projected_log` will include this all updates/deletes (even uncommitted ones) up to this version.
So that later on, in step 3, before starting the pushes, we will update the backfill interval based on the projected log entries.
Can we use this mechanism also for being able to update (uncommitted) new objects such as `obj_k` above?

athanatos replied:
The backfill_interval version should always be set to the last_update_ondisk on the osd where the scan was performed at the time when it started because last_update_ondisk is, definitionally, the version such that all prior updates will be caught by the scan.

athanatos replied:
While progressing through the interval, the primary's interval should then be caught up using the projected log.

athanatos replied:
@xxhdx1985126 In your example above, it's not clerar to me why obj_k wouldn't be added by the log scan.  As long as store_scan + log_scan includes all mutations, that should be enough?

athanatos replied:
Also, classic doesn't hold the pg lock until commit, it should have essentially the same problem.

xxhdx1985126 replied:
> @xxhdx1985126 In your example above, it's not clerar to me why obj_k wouldn't be added by the log scan. As long as store_scan + log_scan includes all mutations, that should be enough?

I think this happens when `store_scan` finishes before `obj_k` is committed. In the current implementation, after `store_scan` and before `log_scan`, we set `primary_bi.version` to `last_update`, which leads to the later `log_scan` leaking `obj_k`

athanatos replied:
If store_scan finishes before obj_k is visible, obj_k should still be picked up from scanning the projected log.

xxhdx1985126 replied:
> If store_scan finishes before obj_k is visible, obj_k should still be picked up from scanning the projected log.

In the current implementation, `primary_bi.version` is set to `last_update` after the `store_scan` and before the `log_scan`, so scanning the projected log won't see `obj_k` as it scan objects the comes after `primary_bi.version`. Am I missing something?

athanatos replied:
Oh, I see.  Yeah, the store_scan should actually set primary_bi.version to the last version committed (checked prior to the scan) rather than last_update because we move entries from the projected_log to the log at write submit time rather than write commit time.  Scanning forward from there to the end of the projected_log would then catch any uncommitted writes.

I *think* last_update_applied is the correct bound to use, but we aren't maintaining it yet in crimson.  pg_committed_to is more conservative, but should work.

Matan-B replied:
Is there any reason for us to actually scan the projected log if so? 
Should we comment it out that until `last_update_applied` is supported?

athanatos replied:
No, we still need to scan the projected log -- otherwise we'd miss objects that are currently being mutated.

Matan-B commented:
jenkins test make check

Matan-B commented:
jenkins test make check

xxhdx1985126 commented:
I'm still testing this PR on our testbeds.

xxhdx1985126 commented:
jenkins test make check

Matan-B commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -162,7 +162,7 @@ void BackfillState::Enqueuing::maybe_update_range()
     logger().debug("{}: scanning projected log", __func__);
     pg().get_projected_log().scan_log_after(primary_bi.version, func);
-    primary_bi.version = pg().get_projected_last_update();
+    primary_bi.version = peering_state().get_pg_committed_to();
```
We have just finished to scan the projected log here, the version should actually be the projected_last_update, right?
I think this case is different than the other version updates in this commit.

xxhdx1985126 replied:
Um, not really, as illustrated in https://github.com/ceph/ceph/pull/59615#discussion_r1825503095, scanning the log can't guarantee that the uncommitted object creation be included properly, so we need to make sure the version of the backfill interval progresses as conservatively as possible.

Matan-B replied:
I would like another pair of eyes here, @athanatos can you please take a look? Thanks!

athanatos replied:
See my comment above.

athanatos replied:
To clarify, after actually scanning the log and projected log, I think that primary_bi.version should be set to projected_last_update.

xxhdx1985126 replied:
Done, please take a look, again, thanks.

Matan-B commented on src/crimson/osd/pg_recovery.cc:
```diff
@@ -533,10 +533,10 @@ void PGRecovery::enqueue_push(
   auto [recovering, added] = pg->get_recovery_backend()->add_recovering(obj);
   if (!added)
     return;
+  peering_state.prepare_backfill_for_missing(obj, v, peers);
```
nit: This commit is not related to the previous commit of this PR, right?

xxhdx1985126 replied:
Um, I think it's related, because an object may be enqueued for push repeatly by BackfillState with the previous commit, so we need to prevent it from making harmful changes.

Matan-B reviewed (commented):
@xxhdx1985126 this one needs a rebase to move to testing

xxhdx1985126 commented:
Rebased.

xxhdx1985126 commented:
rebased.

Matan-B commented:
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-only-testing-13-feb
